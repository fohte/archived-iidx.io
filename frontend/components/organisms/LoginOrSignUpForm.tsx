import * as React from 'react'
import { Field as FinalField, Form as FinalForm } from 'react-final-form'
import {
  Button,
  Form,
  Icon,
  Input,
  Label,
  Message,
  Segment,
} from 'semantic-ui-react'
import { isEmail, isEmpty } from 'validator'

import withSubmitHandling, {
  ComponentExternalProps,
  InjectedProps,
} from '@app/lib/withSubmitHandling'

export interface ExternalProps extends ComponentExternalProps<FormValues> {
  submitText: string
}

export interface Props extends ExternalProps, InjectedProps {}

export interface FormValues {
  email: string
  password: string
}

const validators: {
  [key in keyof FormValues]: (
    value: FormValues[key] | null,
    allValues: FormValues,
  ) => string | undefined
} = {
  email: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your email.'
    }

    if (!isEmail(value)) {
      return 'Email is invalid.'
    }
  },
  password: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your password.'
    }
  },
}

const LoginOrSignUpForm: React.SFC<Props> = ({ handleSubmit, submitText }) => (
  <FinalForm
    onSubmit={handleSubmit}
    initialValues={{ email: '', password: '' }}
  >
    {({
      handleSubmit: innerHandleSubmit,
      pristine,
      hasValidationErrors,
      hasSubmitErrors,
      submitting,
      submitError,
    }) => (
      <Form size="large" onSubmit={innerHandleSubmit} error={hasSubmitErrors}>
        <Segment stacked textAlign="left">
          <FinalField name="email" validate={validators.email}>
            {({ input, meta }) => (
              <Form.Field error={!!(meta.touched && meta.error)}>
                <Input
                  fluid
                  type="email"
                  icon="user"
                  iconPosition="left"
                  {...input}
                  placeholder="E-mail address"
                />
                {meta.touched &&
                  meta.error && (
                    <Label basic color="red" pointing>
                      {meta.error}
                    </Label>
                  )}
              </Form.Field>
            )}
          </FinalField>
          <FinalField name="password" validate={validators.password}>
            {({ input, meta }) => (
              <Form.Field error={!!(meta.touched && meta.error)}>
                <Input
                  fluid
                  type="password"
                  icon="lock"
                  iconPosition="left"
                  {...input}
                  placeholder="Password"
                />
                {meta.touched &&
                  meta.error && (
                    <Label basic color="red" pointing>
                      {meta.error}
                    </Label>
                  )}
              </Form.Field>
            )}
          </FinalField>
          {hasSubmitErrors && <Message error content={submitError} />}
          <Button
            type="submit"
            disabled={submitting || pristine || hasValidationErrors}
            color="teal"
            fluid
            size="large"
          >
            {submitting ? <Icon loading name="spinner" /> : submitText}
          </Button>
        </Segment>
      </Form>
    )}
  </FinalForm>
)

const WrappedNormalLoginOrSignUpForm = withSubmitHandling<FormValues>()(
  LoginOrSignUpForm,
)

export default WrappedNormalLoginOrSignUpForm
