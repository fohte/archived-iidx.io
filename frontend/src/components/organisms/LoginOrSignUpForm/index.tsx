import * as React from 'react'
import { Field as FinalField, Form as FinalForm } from 'react-final-form'
import { isEmail, isEmpty } from 'validator'

import Alert from '@app/components/atoms/Alert'
import Button from '@app/components/atoms/Button'
import FormGroup from '@app/components/atoms/FormGroup'
import InputText from '@app/components/atoms/InputText'
import withSubmitHandling, {
  ComponentExternalProps,
  InjectedProps,
} from '@app/lib/withSubmitHandling'

export interface ExternalProps extends ComponentExternalProps<FormValues> {
  submitText: string
}

export interface Props extends ExternalProps, InjectedProps {}

export interface FormValues {
  email: string
  password: string
}

const validators: {
  [key in keyof FormValues]: (
    value: FormValues[key] | null,
    allValues: FormValues,
  ) => string | undefined
} = {
  email: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your email.'
    }

    if (!isEmail(value)) {
      return 'Email is invalid.'
    }
  },
  password: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your password.'
    }
  },
}

const LoginOrSignUpForm: React.SFC<Props> = ({ handleSubmit, submitText }) => (
  <FinalForm
    onSubmit={handleSubmit}
    initialValues={{ email: '', password: '' }}
  >
    {({
      handleSubmit: innerHandleSubmit,
      pristine,
      hasValidationErrors,
      hasSubmitErrors,
      submitting,
      submitError,
    }) => (
      <form onSubmit={innerHandleSubmit}>
        <FinalField name="email" validate={validators.email}>
          {({ input, meta }) => (
            <FormGroup
              label="E-mail address"
              error={!!(meta.touched && meta.error)}
              errorMessage={meta.error}
            >
              <InputText
                type="email"
                error={!!(meta.touched && meta.error)}
                {...input}
                placeholder="E-mail address"
              />
            </FormGroup>
          )}
        </FinalField>
        <FinalField name="password" validate={validators.password}>
          {({ input, meta }) => (
            <FormGroup
              label="Password"
              error={!!(meta.touched && meta.error)}
              errorMessage={meta.error}
            >
              <InputText
                type="password"
                error={!!(meta.touched && meta.error)}
                {...input}
                placeholder="Password"
              />
            </FormGroup>
          )}
        </FinalField>
        {hasSubmitErrors && <Alert>{submitError}</Alert>}
        <Button
          type="submit"
          disabled={submitting || pristine || hasValidationErrors}
          loading={submitting}
        >
          {submitText}
        </Button>
      </form>
    )}
  </FinalForm>
)

const WrappedNormalLoginOrSignUpForm = withSubmitHandling<FormValues>()(
  LoginOrSignUpForm,
)

export default WrappedNormalLoginOrSignUpForm
