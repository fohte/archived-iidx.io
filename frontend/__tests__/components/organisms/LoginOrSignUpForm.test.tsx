import * as React from 'react'
import * as renderer from 'react-test-renderer'

import LoginOrSignUpForm from '@app/components/organisms/LoginOrSignUpForm'

describe('LoginOrSignUpForm', () => {
  it('renders correctly', () => {
    const component = renderer.create(<LoginOrSignUpForm submitText="Log in" />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
