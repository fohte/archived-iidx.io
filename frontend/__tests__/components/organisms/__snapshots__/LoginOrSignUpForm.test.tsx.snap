// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginOrSignUpForm renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label>
      Email
    </label>
    <input
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
  </div>
  <div>
    <label>
      Password
    </label>
    <input
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <button
    disabled={true}
    type="submit"
  >
    Log in
  </button>
</form>
`;
